### PR TITLE
[SPARK-8233][SQL] misc function: hash

### PR DESCRIPTION
**Pull request by qiansl127**
https://issues.apache.org/jira/browse/SPARK-8233

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 1]
```diff
@@ -47,4 +48,46 @@ case class Md5(child: Expression)
...
+      var res: Int = 0
+      for (i <- 0 until r.length) {
+        res += 31 * res + hashCode(r.get(i))
+      }
+      res
+    }
+    case others => others.hashCode()
+  }
+
+  override def toString: String = s"HASH($child)"
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Remove the `toString` as the case class will handle it properly.

**Comment by chenghao-intel** [comment]
LGTM except a minor issue.
@rxin can you trigger the unit test?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 2]
```diff
@@ -47,4 +48,46 @@ case class Md5(child: Expression)
       "org.apache.spark.unsafe.types.UTF8String.fromString" +
         s"(org.apache.commons.codec.digest.DigestUtils.md5Hex($c))")
   }
+
+}
+
+/**
+ * A function that returns a hash value of the argument
+ */
+case class Hash(child: Expression) extends UnaryExpression {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: In Hive, the `Hash` can take multiple columns: `hash(a1[, a2...])`

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Oh, yes, @qiansl127 can you update this?

Comment (qiansl127) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: @chenghao-intel Done!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 3]
```diff
@@ -47,4 +48,44 @@ case class Md5(child: Expression)
...
+ * A function that returns a hash value of the argument
+ */
+case class Hash(children: Expression*) extends Expression {
+
+  override def foldable: Boolean = children.forall(_.foldable)
+  override def nullable: Boolean = children(0).nullable
+  override def dataType: DataType = IntegerType
+
+  override def eval(input: InternalRow): Any = {
+    val arglist = children.map(_.eval(input).asInstanceOf[AnyRef])
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Could we make argList as InternalRow, then we can use InternalRow.hashCode()?

Thinking of let `hash(A, B, C)` be consistent as `group by A, B, C`, we can know the partition after shuffle by `hash(A, B, C) % numPartitions`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 4]
```diff
@@ -47,4 +48,44 @@ case class Md5(child: Expression)
...
+    case r: Row => {
+      var res: Int = 0
+      for (i <- 0 until r.length) {
+        res += 31 * res + hashCode(r.get(i))
+      }
+      res
+    }
+    case others => others.hashCode()
+  }
+
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Could you add a codegen support?

Comment (qiansl127) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: @davies The codegen support for Hash is so complex but maybe not so useful, can we just make it another PR after some benchmark evaluation work done?

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: OK. If you change to turn them into InternalRow, then it's easy to get hashCode even with codegen.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala (hunk 3 above)
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: I think that's a great idea to make the alignment!
But after looking at the `InternalRow.hashCode()`, seems it doesn't enumerate all of the data types, particularly for the  non primitive type. So are you suggesting to move this implementation to `InternalRow`?

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Is there other types we should take it special? They all fall back to call `hashCode()`

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Like the complex data type.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: @chenghao-intel You are right, we should fix that, especially when there is BinaryType in ArrayType. Could you sending a PR for that?

**Comment by qiansl127** [comment]
@davies Could you check this PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala (hunk 3 above)
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: @davies #7189

**Comment by chenghao-intel** [comment]
@davies, can you please trigger the unit test? I am working on the `hashCode` stulff right now, it can be considered as the improvement, probably this PR would better not to be blocked by the improvement.

**Comment by chenghao-intel** [comment]
And definitely I will update the code here after this PR being merged.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by qiansl127** [comment]
@rxin I've done the rebase, please trigger the test.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala (hunk 4 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: +1 on codegen support. much fewer branches.

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: That's required a recursive revisting all of its elements for the complex data type, code gen here probably be difficult (not easy to maintain compare to the performance gain).

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Actually I am thinking if we need to add wrappers for ALL of the internal data type, which hide the details for the expression developers. e.g the hashCode, serialization / deseriazliation, equality etc. Currently seems the expression developers needs to keep eyes on those details.

Any ideas?

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: It's very expensive to have a wrapper even for primitive types.

**Comment by qiansl127** [comment]
@rxin @davies Could you trigger the test?

**Comment by rxin** [comment]
Jenkins, ok to test.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala (hunk 4 above)
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Yes, so we need to reuse the wrapper, this will be helpful for the expression evaluation, since the result type always be the same, and even, we will not have box/unbox for primitive types.

**Comment by chenghao-intel** [comment]
In general, I think it's ok to leave the `hashCode` for further improvement, for now, I think this PR looks good to me.
Can you guys trigger the unit test?

**Comment by rxin** [comment]
@qiansl127 can you close this pull request?

@JoshRosen will create an implementation based on your change. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 5]
```diff
@@ -47,6 +47,45 @@ case class Md5(child: Expression) extends UnaryExpression with ExpectsInputTypes
...
+    case null => 0
+    case arr: Array[_] => {
+      arr.foldLeft(0) { (acc, n) => acc + hashCode(n) }
+    }
+    case seq: Seq[_] => {
+      seq.foldLeft(0) { (acc, n) => acc + hashCode(n) }
+    }
+    case m: Map[_, _] => {
+      var res: Int = 0
+      m.foreach { case (k, v) => res += (hashCode(k) + hashCode(v)) }
```
Comment (JoshRosen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: The ordering of the map keys in `foreach` is not stable, so I think this may lead to unstable hashcodes for identical maps.

**Comment by qiansl127** [comment]
OK, sorry for no following these days.
